addons/purchase_stock/models/purchase.py

```diff
@@ -103,6 +103,14 @@ def button_cancel(self):
                         if all(state in ('done', 'cancel') for state in siblings_states):
                             move_dest_ids.write({'procure_method': 'make_to_stock'})
                             move_dest_ids._recompute_state()
+                        if order_line.propagate:
+                            move_dest_ids._action_cancel()
+            if order.state in ('purchase'):
+                for order_line in order.order_line:
+                    if order_line.move_ids:
+                        move_ids = order_line.move_ids.filtered(lambda m: m.state not in ('done', 'cancel'))
```

Comment:
filtered not needed you already checked before if the moves was not done and action_cancel skip cancelled moves